bulldozer/evaluate.go

```diff
@@ -28,130 +27,23 @@ import (
 // IsPRBlacklisted returns true if the PR is identified as blacklisted,
 // false otherwise. Additionally, a description of the reason will be returned.
 func IsPRBlacklisted(ctx context.Context, pullCtx pull.Context, config Signals) (bool, string, error) {
-	labels, err := pullCtx.Labels(ctx)
+	matches, reason, err := config.Matches(ctx, pullCtx, "blacklist")
 	if err != nil {
-		return true, "unable to list PR labels", err
+		// blacklist must always fail closed (matches on error)
+		matches = true
 	}
-
-	if inSlice, idx := anyInSliceCaseInsensitive(labels, config.Labels); inSlice {
-		return true, fmt.Sprintf("PR label matches one of specified blacklist labels: %q", config.Labels[idx]), nil
-	}
-
-	body, err := pullCtx.Body(ctx)
-	if err != nil {
-		return true, "unable to list PR body", err
-	}
-
-	comments, err := pullCtx.Comments(ctx)
-	if err != nil {
-		return true, "unable to list PR comments", err
-	}
-
-	if inSlice, idx := anyInSlice(comments, config.Comments); inSlice {
-		return true, fmt.Sprintf("PR comment matches one of specified blacklist comments: %q", config.Comments[idx]), nil
-	}
-
-	for _, blacklistedComment := range config.Comments {
-		if blacklistedComment == body {
-			return true, fmt.Sprintf("PR body matches one of specified blacklist comments: %q", blacklistedComment), nil
-		}
-	}
-
-	for _, blacklistedSubstring := range config.CommentSubstrings {
-		for _, comment := range comments {
-			if strings.Contains(comment, blacklistedSubstring) {
-				return true, fmt.Sprintf("PR comment matches one of specified blacklist comment substrings: %q", blacklistedSubstring), nil
-			}
-		}
-
-		if strings.Contains(body, blacklistedSubstring) {
-			return true, fmt.Sprintf("PR body matches one of specified blacklist comment substrings: %q", blacklistedSubstring), nil
-		}
-	}
-
-	for _, blacklistedSubstring := range config.PRBodySubstrings {
-		if strings.Contains(body, blacklistedSubstring) {
-			return true, fmt.Sprintf("PR body matches one of specified blacklist substrings: %q", blacklistedSubstring), nil
-		}
-	}
-
-	return false, "no matching blacklist found", nil
+	return matches, reason, err
 }
 
 // IsPRWhitelisted returns true if the PR is identified as whitelisted,
 // false otherwise. Additionally, a description of the reason will be returned.
 func IsPRWhitelisted(ctx context.Context, pullCtx pull.Context, config Signals) (bool, string, error) {
-	labels, err := pullCtx.Labels(ctx)
+	matches, reason, err := config.Matches(ctx, pullCtx, "whitelist")
 	if err != nil {
-		return false, "unable to list PR labels", err
-	}
-
-	if inSlice, idx := anyInSliceCaseInsensitive(labels, config.Labels); inSlice {
-		return true, fmt.Sprintf("PR label matches one of specified whitelist labels: %q", config.Labels[idx]), nil
+		// whitelist must always fail closed (no match on error)
```

Comment:
curious: any way to enforce this via test? (same for blacklist)

Reply:
Yeah, there are already tests for this: https://github.com/palantir/bulldozer/blob/bkeyes/branch-whitelist/bulldozer/evaluate_test.go#L48